File: 85-maximal-rectangle/maximal-rectangle.cpp
class Solution {
public:
    int maximalRectangle(vector<vector<char>>& matrix) {
        int n = matrix.size();
        int m = matrix[0].size();
        vector<vector<int>> v(n,vector<int>(m,0));
         for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
               if(matrix[i][j]=='1')
               {
                v[i][j]=1;
               }
             
            }
        }
        for (int i = 1; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if(v[i][j]!=0){
                v[i][j] += v[i - 1][j];}
            }
        }
        int ans=0;
        for (int i = 0; i < n; i++) {
           v[i].push_back(0);
             stack<int> s;
            for(int j = 0; j < v[i].size(); j++) {
                while(!s.empty() && v[i][s.top()] > v[i][j]) {
                    int height = v[i][s.top()];
                    s.pop();
                    int width = s.empty() ? j : j - s.top() - 1;
                    ans = max(ans, height * width);
                }
                s.push(j);
            }
        }
        return ans;
    }
};

File: 85-maximal-rectangle/maximal-rectangle.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        int n = heights.length;
        Stack<int[]> st = new Stack<>();
        int ans=0;  
        for(int i=0;i<n;i++)
        {
            int ind=i;
            while(!st.isEmpty() && st.peek()[1]>heights[i])
            {
                int[] x = st.pop();
                ind=Math.min(ind,x[0]);
                ans=Math.max(ans,x[1]*(i-x[0]));
                ans=Math.max(ans,x[1]);
            }
            st.push(new int[]{ind,heights[i]});
        }
        while(!st.isEmpty())
        {
            int[] x=st.pop();
            ans=Math.max(ans,x[1]*(n-x[0]));
        }
        return ans;
    }
    public int maximalRectangle(char[][] matrix) {
        int n = matrix.length;
        int m = matrix[0].length;
        int[][] depth = new int[n][m];
       
        int ans = 0;
       
        for (int j = 0; j < m; j++) {
            int c = 0;
            for (int i = n - 1; i >= 0; i--) {
                if (matrix[i][j] == '1') {
                    c++;
                } else {
                    c = 0;
                }
                depth[i][j] = c;
            }
        }
       
        for(int i=0;i<n;i++)
        {
            ans=Math.max(ans,largestRectangleArea(depth[i]));
        }

        return ans;
    }
}